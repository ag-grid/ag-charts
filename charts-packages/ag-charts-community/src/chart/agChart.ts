import { Chart } from "./chart";
import { Series } from "./series/series";
import { ChartAxis } from "./chartAxis";
import { find } from "../util/array";
import { LegendMarker } from "./legend";
import { ChartTheme, ChartPalette, IChartTheme } from "./themes/chartTheme";
import { DarkTheme } from './themes/darkTheme';
import { getValue } from "../util/object";
import mappings from './chartMappings';
import { MaterialLight } from "./themes/materialLight";
import { MaterialDark } from "./themes/materialDark";
import { PastelLight } from "./themes/pastelLight";
import { PastelDark } from "./themes/pastelDark";
import { SolarLight } from "./themes/solarLight";
import { SolarDark } from "./themes/solarDark";
import { VividLight } from "./themes/vividLight";
import { VividDark } from "./themes/vividDark";

const lightTheme = new ChartTheme();
const themes = {
    default: lightTheme,
    light: lightTheme,
    dark: new DarkTheme(),
    'material-light': new MaterialLight(),
    'material-dark': new MaterialDark(),
    'pastel-light': new PastelLight(),
    'pastel-dark': new PastelDark(),
    'solar-light': new SolarLight(),
    'solar-dark': new SolarDark(),
    'vivid-light': new VividLight(),
    'vivid-dark': new VividDark()
} as any;

function getTheme(value: string | ChartTheme | IChartTheme): ChartTheme {
    if (typeof value === 'string') {
        return themes[value] || themes.default;
    }
    if (value instanceof ChartTheme) {
        return value;
    }
    if (value && (value.defaults || value.palette)) {
        const baseTheme: any = getTheme(value.baseTheme);
        return new baseTheme.constructor(value);
    }
    return themes.default;
}

export abstract class AgChart {
    static create(options: any, container?: HTMLElement, data?: any[]) {
        options = Object.create(options); // avoid mutating user provided options
        if (container) {
            options.container = container;
        }
        if (data) {
            options.data = data;
        }
        // special handling when both `autoSize` and `width` / `height` are present in the options
        const autoSize = options && options.autoSize;
        const theme = getTheme(options.theme);
        const chart = create(options, undefined, undefined, theme);
        if (chart) {
            if (autoSize) {  // `autoSize` takes precedence over `width` / `height`
                chart.autoSize = true;
            }
            if (theme) {
                theme.updateChart(chart);
            }
        }
        // console.log(JSON.stringify(flattenObject(options), null, 4));
        return chart;
    }

    static update(chart: any, options: any, container?: HTMLElement, data?: any[]) {
        options = Object.create(options);
        if (container) {
            options.container = container;
        }
        if (data) {
            options.data = data;
        }
        const autoSize = options && options.autoSize;
        const theme = getTheme(options.theme);
        update(chart, options, undefined, theme);
        if (chart) {
            if (autoSize) {
                chart.autoSize = true;
            }
            if (theme) {
                theme.updateChart(chart);
            }
        }
    }

    static createComponent = create;
}

const pathToSeriesTypeMap: { [key in string]: string } = {
    'cartesian.series': 'line', // default series type for cartesian charts
    'line.series': 'line',
    'area.series': 'area',
    'bar.series': 'bar',
    'column.series': 'column',
    'scatter.series': 'scatter',
    'polar.series': 'pie', // default series type for polar charts
    'pie.series': 'pie'
};

function create(options: any, path?: string, component?: any, theme?: ChartTheme) {
    // Deprecate `chart.legend.item.marker.type` in integrated chart options.
    options = Object.create(options);
    if (component instanceof LegendMarker) {
        if (options.type) {
            options.shape = options.type;
        }
    } else {
        options = provideDefaultType(options, path);
        if (path) {
            if (options.type) {
                path = path + '.' + options.type;
            }
        } else {
            path = options.type;
        }
    }

    const mapping = getValue(mappings, path);

    if (mapping) {
        options = provideDefaultOptions(options, mapping, theme && theme.getConfig(path));

        const meta = mapping.meta || {};
        const constructorParams = meta.constructorParams || [];
        const skipKeys = ['type', 'listeners'].concat(constructorParams);
        // TODO: Constructor params processing could be improved, but it's good enough for current params.
        const constructorParamValues = constructorParams
            .map((param: any) => options[param])
            .filter((value: any) => value !== undefined);

        component = component || new meta.constructor(...constructorParamValues);

        for (const key in options) {
            // Process every non-special key in the config object.
            if (skipKeys.indexOf(key) < 0) {
                const value = options[key];

                if (value && key in mapping && !(meta.setAsIs && meta.setAsIs.indexOf(key) >= 0)) {
                    if (Array.isArray(value)) {
                        const subComponents = value.map(config => create(config, path + '.' + key, undefined, theme)).filter(config => !!config);
                        component[key] = subComponents;
                    } else {
                        if (mapping[key] && component[key]) {
                            // The instance property already exists on the component (e.g. chart.legend).
                            // Simply configure the existing instance, without creating a new one.
                            create(value, path + '.' + key, component[key], theme);
                        } else {
                            const subComponent = create(value, value.type ? path : path + '.' + key, undefined, theme);
                            if (subComponent) {
                                component[key] = subComponent;
                            }
                        }
                    }
                } else { // if (key in meta.constructor.defaults) { // prevent users from creating custom properties
                    component[key] = value;
                }
            }
        }

        const listeners = options.listeners;
        if (component && component.addEventListener && listeners) {
            for (const key in listeners) {
                if (listeners.hasOwnProperty(key)) {
                    const listener = listeners[key];
                    if (typeof listener === 'function') {
                        component.addEventListener(key, listener);
                    }
                }
            }
        }

        return component;
    }
}

function update(component: any, options: any, path?: string, theme?: ChartTheme) {
    if (!(options && typeof options === 'object')) {
        return;
    }

    // Deprecate `chart.legend.item.marker.type` in integrated chart options.
    if (component instanceof LegendMarker) {
        if (options.type) {
            options.shape = options.type;
        }
    } else {
        options = provideDefaultType(options, path);
        if (path) {
            if (options.type) {
                path = path + '.' + options.type;
            }
        } else {
            path = options.type;
        }
    }

    const mapping = getValue(mappings, path);

    if (mapping) {
        options = provideDefaultOptions(options, mapping, theme && theme.getConfig(path));

        const meta = mapping.meta || {};
        const constructorParams = meta && meta.constructorParams || [];
        const skipKeys = ['type'].concat(constructorParams);

        for (const key in options) {
            if (skipKeys.indexOf(key) < 0) {
                const value = options[key];
                const keyPath = path + '.' + key;

                if (meta.setAsIs && meta.setAsIs.indexOf(key) >= 0) {
                    component[key] = value;
                } else {
                    const oldValue = component[key];

                    if (Array.isArray(oldValue) && Array.isArray(value)) {
                        if (path in mappings) { // component is a chart
                            if (key === 'series') {
                                updateSeries(component, value, keyPath, theme);
                            } else if (key === 'axes') {
                                updateAxes(component, value, keyPath, theme);
                            }
                        } else {
                            component[key] = value;
                        }
                    } else if (typeof oldValue === 'object') {
                        if (value) {
                            update(oldValue, value, value.type ? path : keyPath, theme);
                        } else if (key in options) {
                            component[key] = value;
                        }
                    } else {
                        const subComponent = isObject(value) && create(value, value.type ? path : keyPath, undefined, theme);
                        if (subComponent) {
                            component[key] = subComponent;
                        } else {
                            component[key] = value;
                        }
                    }
                }
            }
        }
    }

    if (path in mappings) { // top-level component (chart)
        (component as Chart).performLayout();
    }
}

function updateSeries(chart: Chart, configs: any[], keyPath: string, theme?: ChartTheme) {
    const allSeries = chart.series;
    let prevSeries: Series | undefined;
    let i = 0;
    for (; i < configs.length; i++) {
        let config = configs[i];
        let series = allSeries[i];
        if (series) {
            config = provideDefaultType(config, keyPath);
            if (series.type === config.type) {
                update(series, config, keyPath, theme);
            } else {
                const newSeries = create(config, keyPath, undefined, theme);
                chart.removeSeries(series);
                chart.addSeriesAfter(newSeries, prevSeries);
                series = newSeries;
            }
        } else { // more new configs than existing series
            const newSeries = create(config, keyPath, undefined, theme);
            chart.addSeries(newSeries);
        }
        prevSeries = series;
    }
    // more existing series than new configs
    for (; i < allSeries.length; i++) {
        const series = allSeries[i];
        if (series) {
            chart.removeSeries(series);
        }
    }
}

function updateAxes(chart: Chart, configs: any[], keyPath: string, theme?: ChartTheme) {
    const axes = chart.axes as ChartAxis[];
    const axesToAdd: ChartAxis[] = [];
    const axesToUpdate: ChartAxis[] = [];

    for (const config of configs) {
        const axisToUpdate = find(axes, axis => {
            return axis.type === config.type && axis.position === config.position;
        });
        if (axisToUpdate) {
            axesToUpdate.push(axisToUpdate);
            update(axisToUpdate, config, keyPath, theme);
        } else {
            const axisToAdd = create(config, keyPath, undefined, theme);
            if (axisToAdd) {
                axesToAdd.push(axisToAdd);
            }
        }
    }

    chart.axes = axesToUpdate.concat(axesToAdd);
}

function provideDefaultChartType(options: any): any {
    if (options.type) {
        return options;
    }
    // If chart type is not specified, try to infer it from the type of first series.
    const series = options.series && options.series[0];

    if (series && series.type) {
        outerLoop: for (const chartType in mappings) {
            for (const seriesType in mappings[chartType].series) {
                if (series.type === seriesType) {
                    options = Object.create(options);
                    options.type = chartType;
                    break outerLoop;
                }
            }
        }
    }
    if (!options.type) {
        options = Object.create(options);
        options.type = 'cartesian';
    }

    return options;
}

function provideDefaultType(options: any, path?: string): any {
    if (!path) { // if `path` is undefined, `options` is a top-level (chart) config
        options = provideDefaultChartType(options);
    }

    if (!options.type) {
        const seriesType = pathToSeriesTypeMap[path];
        if (seriesType) {
            options = Object.create(options);
            options.type = seriesType;
        }
    }

    return options;
}

/**
 * If certain options were not provided by the user, use the defaults from the theme and the mapping.
 * All three objects are provided for the current path in the config tree, not necessarily top-level.
 * @param options
 * @param mapping
 * @param themeDefaults
 */
function provideDefaultOptions(options: any, mapping: any, themeDefaults?: any): any {
    const defaults = mapping && mapping.meta && mapping.meta.defaults;

    if (defaults || themeDefaults) {
        options = Object.create(options);
    }
    // Fill in the gaps for properties not configured by the user using theme provided values.
    for (const key in themeDefaults) {
        if (!(key in options)) {
            options[key] = themeDefaults[key];
        }
    }
    // Fill in the gaps for properties not configured by the user, nor theme using chart mappings.
    for (const key in defaults) {
        if ((!themeDefaults || !(key in themeDefaults)) && !(key in options)) {
            options[key] = defaults[key];
        }
    }

    return options;
}

function isObject(value: any): boolean {
    return typeof value === 'object' && !Array.isArray(value);
}
