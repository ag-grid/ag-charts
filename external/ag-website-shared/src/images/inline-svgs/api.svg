<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2" viewBox="0 0 32 32"><path d="M24 22v-1h6v-2h-6v-6h6v-2h-6v-1c0-1.097-.903-2-2-2h-6c-4.009.007-7.426 3.023-7.931 7H2v2h6.069c.505 3.977 3.922 6.993 7.931 7h6c1.097 0 2-.903 2-2Zm-8 0c-3.292 0-6-2.708-6-6s2.708-6 6-6h6v12h-6Z" style="fill-rule:nonzero"/></svg>