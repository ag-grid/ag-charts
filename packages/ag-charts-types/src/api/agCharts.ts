import type { AgChartInstance, AgChartOptions, AgFinancialChartOptions, AgGaugeOptions } from '../chartBuilderOptions';

export interface AgChartsApi {
    /** Create a new `AgChartInstance` based upon the given configuration options. */
    create<T extends AgChartOptions>(options: AgChartOptions): AgChartInstance<T>;
    /** Create a new `AgChartInstance` based upon the given configuration options. */
    createFinancialChart(options: AgFinancialChartOptions): AgChartInstance<AgFinancialChartOptions>;
    /** Create a new `AgChartInstance` based upon the given configuration options. */
    createGauge(options: AgGaugeOptions): AgChartInstance<AgGaugeOptions>;

    setLicenseKey(licenseKey: string): void;

    /** @private Internal to AG Grid, returns the `AgChartInstance` for a DOM node, if there is one. */
    getInstance(element: HTMLElement): AgChartInstance | undefined;
    /** @private Internal to AG Grid, not for end-user consumption. */
    setGridContext(enabled: boolean): void;
}
