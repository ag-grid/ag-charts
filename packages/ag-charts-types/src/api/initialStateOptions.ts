import type { AgAnnotation } from '../chart/annotationsOptions';

// import type { AgStateSerializableDate } from './stateTypes';

export interface AgInitialStateOptions {
    /** The initial set of annotations to display on the chart. */
    annotations?: AgAnnotation[];
    /** The initial zoom state. */
    // zoom?: AgInitialStateZoomOptions;
}

// export interface AgInitialStateZoomOptions {
//     rangeX?: AgInitialStateZoomRange;
//     rangeY?: AgInitialStateZoomRange;
//     ratioX?: AgInitialStateZoomRatio;
//     ratioY?: AgInitialStateZoomRatio;
// }

// export interface AgInitialStateZoomRange {
//     start?: AgStateSerializableDate | number;
//     end?: AgStateSerializableDate | number;
// }

// export interface AgInitialStateZoomRatio {
//     start?: number;
//     end?: number;
// }
