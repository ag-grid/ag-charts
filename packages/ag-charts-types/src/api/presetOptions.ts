import type { AgLinearGaugeSeriesOptions } from '../series/gauge/linearGaugeOptions';
import type { AgRadialGaugeSeriesOptions } from '../series/gauge/radialGaugeOptions';

/** @deprecated v10.1.0 use 'hlc' instead. */
export type DeprecatedAgPriceVolumeChartType = 'range-area';
export type AgPriceVolumeChartType =
    | 'candlestick'
    | 'hollow-candlestick'
    | 'ohlc'
    | 'line'
    | 'step-line'
    | 'hlc'
    | 'high-low'
    | DeprecatedAgPriceVolumeChartType;

export interface AgPriceVolumePreset {
    /** Series type used for the OHLC data.
     *
     *  Default: `'candlestick'`
     */
    chartType?: AgPriceVolumeChartType;
    /** The key used to retrieve x-values from the data.
     *
     * Default: `'date'`
     *
     * @deprecated v10.1.0 use `dateKey` instead.
     */
    xKey?: string;
    /** The key used to retrieve x-values from the data.
     *
     * Default: `'date'`
     */
    dateKey?: string;
    /** The key used to retrieve 'open' values from the data.
     *
     * Default: `'open'`
     */
    openKey?: string;
    /** The key used to retrieve 'high' values from the data.
     *
     * Default: `'high'`
     */
    highKey?: string;
    /** The key used to retrieve 'low' values from the data.
     *
     * Default: `'low'`
     */
    lowKey?: string;
    /** The key used to retrieve 'close' values from the data.
     *
     *  Default: `'close'`
     */
    closeKey?: string;
    /** The key used to retrieve 'volume' values from the data.
     *
     * Default: `'volume'`
     */
    volumeKey?: string;
    /** Whether to show the Navigator and mini-chart beneath the main chart.
     *
     * Default: `false`
     */
    navigator?: boolean;
    /** Whether to show the volume series at the bottom of the chart.
     *
     *  If set to `false`, no volume data is required.
     *
     * Default: `true`
     */
    volume?: boolean;
    /** Whether to show the range toolbar.
     *
     * Default: `true`
     *
     * @deprecated v10.1.0 use `rangeButtons` instead.
     */
    rangeToolbar?: boolean;
    /** Whether to show the range buttons.
     *
     * Default: `true`
     */
    rangeButtons?: boolean;
    /** Whether to show the status bar.
     *
     * Default: `true`
     */
    statusBar?: boolean;
    /** Whether Annotations are enabled.
     *
     * Default: `true`
     *
     * @deprecated v10.1.0 use `toolbar` instead.
     */
    annotations?: boolean;
    /** Whether the toolbar is enabled.
     *
     * Default: `true`
     */
    toolbar?: boolean;
    /** Whether Zoom is enabled.
     *
     * Default: `true`
     */
    zoom?: boolean;
}

export type AgFinancialChartPresets = AgPriceVolumePreset;

export type AgLinearGaugePreset = AgLinearGaugeSeriesOptions;
export type AgRadialGaugePreset = AgRadialGaugeSeriesOptions;
export type AgGaugePreset = AgLinearGaugePreset | AgRadialGaugePreset;

export type Preset = AgFinancialChartPresets;
