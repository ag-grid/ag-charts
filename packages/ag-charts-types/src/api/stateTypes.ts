export interface AgStateSerializableDate {
    __type: 'date';
    value: string | number;
}
