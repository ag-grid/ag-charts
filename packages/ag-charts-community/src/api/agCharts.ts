import type { AgChartInstance, AgChartOptions, AgFinancialChartOptions, AgGaugeOptions } from 'ag-charts-types';

import { CartesianChart } from '../chart/cartesianChart';
import { Chart, type ChartExtendedOptions } from '../chart/chart';
import { AgChartInstanceProxy, type FactoryApi } from '../chart/chartProxy';
import { registerInbuiltModules } from '../chart/factory/registerInbuiltModules';
import { setupModules } from '../chart/factory/setupModules';
import { FlowProportionChart } from '../chart/flowProportionChart';
import { GaugeChart } from '../chart/gaugeChart';
import { HierarchyChart } from '../chart/hierarchyChart';
import {
    isAgCartesianChartOptions,
    isAgFlowProportionChartOptions,
    isAgGaugeChartOptions,
    isAgHierarchyChartOptions,
    isAgPolarChartOptions,
    isAgTopologyChartOptions,
} from '../chart/mapping/types';
import { PolarChart } from '../chart/polarChart';
import { TopologyChart } from '../chart/topologyChart';
import type { LicenseManager } from '../module/enterpriseModule';
import { enterpriseModule } from '../module/enterpriseModule';
import { ChartOptions } from '../module/optionsModule';
import { Debug } from '../util/debug';
import { deepClone, jsonWalk } from '../util/json';
import { mergeDefaults } from '../util/object';
import type { DeepPartial } from '../util/types';
import { VERSION } from '../version';
import { MementoCaretaker } from './state/memento';

const debug = Debug.create(true, 'opts');

function chartType(options: any): 'cartesian' | 'polar' | 'hierarchy' | 'topology' | 'flow-proportion' | 'gauge' {
    if (isAgCartesianChartOptions(options)) {
        return 'cartesian';
    } else if (isAgPolarChartOptions(options)) {
        return 'polar';
    } else if (isAgHierarchyChartOptions(options)) {
        return 'hierarchy';
    } else if (isAgTopologyChartOptions(options)) {
        return 'topology';
    } else if (isAgFlowProportionChartOptions(options)) {
        return 'flow-proportion';
    } else if (isAgGaugeChartOptions(options)) {
        return 'gauge';
    }

    throw new Error(`AG Chart - unknown type of chart for options with type: ${options.type}`);
}

/**
 * Factory for creating and updating instances of AgChartInstance.
 *
 * @docsInterface
 */
export abstract class AgCharts {
    private static licenseManager?: LicenseManager;
    private static licenseChecked = false;
    private static licenseKey?: string;
    private static gridContext = false;

    private static licenseCheck(options: AgChartOptions) {
        if (this.licenseChecked) return;

        this.licenseManager = enterpriseModule.licenseManager?.(options);
        this.licenseManager?.setLicenseKey(this.licenseKey, this.gridContext);
        this.licenseManager?.validateLicense();
        this.licenseChecked = true;
    }

    /** @private - for use by Charts website dark-mode support. */
    static optionsMutationFn?: (opts: AgChartOptions, preset?: string) => AgChartOptions;

    public static setLicenseKey(licenseKey: string) {
        this.licenseKey = licenseKey;
    }

    public static setGridContext(gridContext: boolean) {
        this.gridContext = gridContext;
    }

    public static getLicenseDetails(licenseKey: string) {
        return enterpriseModule.licenseManager?.({}).getLicenseDetails(licenseKey);
    }

    /**
     * Returns the `AgChartInstance` for a DOM node, if there is one.
     */
    public static getInstance(element: HTMLElement): AgChartInstance | undefined {
        return AgChartsInternal.getInstance(element);
    }

    /**
     * Create a new `AgChartInstance` based upon the given configuration options.
     */
    public static create<O extends AgChartOptions>(options: O): AgChartInstance<O> {
        this.licenseCheck(options);
        const chart = AgChartsInternal.createOrUpdate(
            options,
            undefined,
            this.licenseManager,
            enterpriseModule.styles != null ? [['ag-charts-enterprise', enterpriseModule.styles]] : []
        );

        if (this.licenseManager?.isDisplayWatermark() && this.licenseManager) {
            enterpriseModule.injectWatermark?.(chart.chart.ctx.domManager, this.licenseManager.getWatermarkMessage());
        }
        return chart as unknown as AgChartInstance<O>;
    }

    public static createFinancialChart(options: AgFinancialChartOptions): AgChartInstance<AgFinancialChartOptions> {
        return this.create({ presetType: 'price-volume', ...options } as AgChartOptions) as any;
    }

    public static createGauge(options: AgGaugeOptions): AgChartInstance<AgGaugeOptions> {
        return this.create({ presetType: 'gauge', ...(options as any) } as AgChartOptions) as any;
    }
}

class AgChartsInternal {
    private static readonly caretaker = new MementoCaretaker(VERSION);

    static getInstance(element: HTMLElement): AgChartInstanceProxy | undefined {
        const chart = Chart.getInstance(element);
        return chart ? AgChartInstanceProxy.chartInstances.get(chart) : undefined;
    }

    private static initialised = false;
    static initialiseModules() {
        if (AgChartsInternal.initialised) return;

        registerInbuiltModules();
        setupModules();

        AgChartsInternal.initialised = true;
    }

    static callbackApi: FactoryApi = {
        caretaker: AgChartsInternal.caretaker,
        createOrUpdate(opts, chart) {
            return AgChartsInternal.createOrUpdate(opts, chart as AgChartInstanceProxy);
        },
        updateUserDelta(chart, deltaOptions) {
            return AgChartsInternal.updateUserDelta(chart as AgChartInstanceProxy, deltaOptions);
        },
    };

    static createOrUpdate(
        options: ChartExtendedOptions,
        proxy?: AgChartInstanceProxy,
        licenseManager?: LicenseManager,
        styles?: Array<[string, string]>
    ) {
        AgChartsInternal.initialiseModules();

        debug('>>> AgCharts.createOrUpdate() user options', options);

        const { presetType = proxy?.chart.chartOptions.presetType, ...otherOptions } = options;

        let mutableOptions = otherOptions;
        if (AgCharts.optionsMutationFn) {
            mutableOptions = AgCharts.optionsMutationFn(mutableOptions, presetType);
            debug('>>> AgCharts.createOrUpdate() MUTATED user options', options);
        }

        const { overrideDevicePixelRatio, document, window: userWindow, ...userOptions } = mutableOptions;
        const chartOptions = new ChartOptions(userOptions, {
            presetType,
            document,
            window: userWindow,
            overrideDevicePixelRatio,
        });

        let chart = proxy?.chart;
        if (chart == null || chartType(userOptions) !== chartType(chart?.chartOptions.processedOptions)) {
            chart = AgChartsInternal.createChartInstance(chartOptions, chart);
            styles?.forEach(([id, css]) => {
                chart?.ctx.domManager.addStyles(id, css);
            });
        }

        if (proxy == null) {
            proxy = new AgChartInstanceProxy(chart, AgChartsInternal.callbackApi, licenseManager);
        } else {
            proxy.chart = chart;
        }

        if (debug.check() && typeof window !== 'undefined') {
            (window as any).agChartInstances ??= {};
            (window as any).agChartInstances[chart.id] = chart;
        }

        chart.queuedUserOptions.push(userOptions);
        chart.requestFactoryUpdate((chartRef) => {
            chartRef.applyOptions(chartOptions);
            // If there are a lot of update calls, `requestFactoryUpdate()` may skip callbacks,
            // so we need to remove all queue items up to the last successfully applied item.
            const queueIdx = chartRef.queuedUserOptions.indexOf(userOptions) + 1;
            chartRef.queuedUserOptions.splice(0, queueIdx);
        });

        return proxy;
    }

    static updateUserDelta(proxy: AgChartInstanceProxy, deltaOptions: DeepPartial<AgChartOptions>) {
        deltaOptions = deepClone(deltaOptions, { shallow: ['data'] });

        jsonWalk(
            deltaOptions,
            (node) => {
                if (typeof node !== 'object') return;
                for (const [key, value] of Object.entries(node)) {
                    if (typeof value === 'undefined') {
                        Object.assign(node, { [key]: Symbol('UNSET') });
                    }
                }
            },
            { skip: ['data'] }
        );

        const { chart } = proxy;
        const lastUpdateOptions = chart.getOptions();
        const userOptions = mergeDefaults(deltaOptions, lastUpdateOptions);
        debug('>>> AgCharts.updateUserDelta() user delta', deltaOptions);
        debug('AgCharts.updateUserDelta() - base options', lastUpdateOptions);
        AgChartsInternal.createOrUpdate(userOptions, proxy);
    }

    private static createChartInstance(options: ChartOptions, oldChart?: Chart): Chart {
        const transferableResource = oldChart?.destroy({ keepTransferableResources: true });
        const ChartConstructor = AgChartsInternal.getChartByOptions(options.processedOptions);
        return new ChartConstructor(options, transferableResource);
    }

    private static getChartByOptions(options: AgChartOptions) {
        if (isAgCartesianChartOptions(options)) {
            return CartesianChart;
        } else if (isAgHierarchyChartOptions(options)) {
            return HierarchyChart;
        } else if (isAgPolarChartOptions(options)) {
            return PolarChart;
        } else if (isAgTopologyChartOptions(options)) {
            return TopologyChart;
        } else if (isAgFlowProportionChartOptions(options)) {
            return FlowProportionChart;
        } else if (isAgGaugeChartOptions(options)) {
            return GaugeChart;
        }

        throw new Error(
            `AG Charts - couldn't apply configuration, check options are correctly structured and series types are specified`
        );
    }
}
