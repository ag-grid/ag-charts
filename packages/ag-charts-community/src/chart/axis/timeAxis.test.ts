import { afterEach, describe, expect, it } from '@jest/globals';

import { AgCharts } from '../../api/agChart';
import type { AgCartesianChartOptions } from '../../options/agChartOptions';
import type { ChartAxis } from '../chartAxis';
import {
    IMAGE_SNAPSHOT_DEFAULTS,
    deproxy,
    extractImageData,
    prepareTestOptions,
    setupMockCanvas,
    waitForChartStability,
} from '../test/utils';

function calculateAxisBBox(axis: ChartAxis): { x: number; y: number; width: number; height: number } {
    const bbox = axis.computeBBox();

    const { x, y, width, height } = bbox;
    return { x, y, width, height };
}

const TIME_AXIS_DATA = [];
for (let i = 0; i < 200; i++) {
    TIME_AXIS_DATA.push({
        date: new Date(2004, 3 + i, 1),
        'Tate Modern': 82215 + (i % 5) * 10_000,
    });
}

const TIME_AXIS_EXAMPLE: AgCartesianChartOptions = {
    container: document.getElementById('myChart'),
    title: {
        text: 'Navigator Styling',
    },
    data: TIME_AXIS_DATA,
    series: [
        {
            type: 'bar',
            xKey: 'date',
            yKey: 'Tate Modern',
            fill: '#c16068',
            stroke: '#874349',
        },
    ],
    axes: [
        {
            type: 'time',
            position: 'bottom',
            nice: false,
            tick: { maxSpacing: 150 },
        },
        {
            type: 'number',
            position: 'left',
            label: {
                formatter: (params) => params.value / 1000 + 'k',
            },
        },
    ],
    navigator: {
        height: 50,
        min: 0,
        max: 1,
    },
};

describe('Time Axis Examples', () => {
    let chart: any;
    const ctx = setupMockCanvas();

    const axisCompare = async () => {
        await waitForChartStability(chart);

        for (const axis of deproxy(chart).axes) {
            if (axis.type !== 'time') continue;

            const axisBbox = calculateAxisBBox(axis);
            expect(axisBbox).toMatchSnapshot();

            const { width, height } = axisBbox;

            if (width > 0 && height > 0) {
                const imageData = extractImageData({ ...ctx, bbox: axisBbox });
                expect(imageData).toMatchImageSnapshot(IMAGE_SNAPSHOT_DEFAULTS);
            }
        }
    };

    afterEach(() => {
        if (chart) {
            chart.destroy();
            (chart as unknown) = undefined;
        }
    });

    describe('when zooming', () => {
        const ZOOM_LEVELS: [number, number][] = [
            [0, 1],
            [0.2, 0.5],
            [0.25, 0.35],
            [0.3, 0.35],
            [0.35, 0.375],
            [0.4, 0.4125],
            [0.45, 0.45625],
        ];
        for (const [min, max] of ZOOM_LEVELS) {
            it(`for should render as expected as zoom [${min}, ${max}]`, async () => {
                chart = AgCharts.create(prepareTestOptions({ ...TIME_AXIS_EXAMPLE }));
                chart.updateDelta({ navigator: { min, max } });
                await axisCompare();
            });
        }
    });
});
