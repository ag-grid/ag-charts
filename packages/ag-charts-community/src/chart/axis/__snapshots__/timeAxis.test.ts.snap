// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time Axis Examples when zooming for should render as expected as zoom [0, 1] 1`] = `
{
  "height": 20.33333333333337,
  "width": 721,
  "x": 59,
  "y": 500,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.2, 0.5] 1`] = `
{
  "height": 20.33333333333337,
  "width": 721,
  "x": 59,
  "y": 500,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.3, 0.35] 1`] = `
{
  "height": 23,
  "width": 725.6666666666667,
  "x": 59,
  "y": 497,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.4, 0.4125] 1`] = `
{
  "height": 23,
  "width": 721,
  "x": 59,
  "y": 497,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.25, 0.35] 1`] = `
{
  "height": 23,
  "width": 721,
  "x": 59,
  "y": 497,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.35, 0.375] 1`] = `
{
  "height": 23,
  "width": 721,
  "x": 59,
  "y": 497,
}
`;

exports[`Time Axis Examples when zooming for should render as expected as zoom [0.45, 0.45625] 1`] = `
{
  "height": 0,
  "width": 721,
  "x": 59,
  "y": 520,
}
`;
